File: analytical.R
library(tidyverse)

#read data
account <- read.csv('data/accounts.csv')
card <- read.csv('data/cards.csv')
district <- read.csv('data/districts.csv')
link <- read.csv('data/links.csv')
loan <- read.csv('Result/loans_r.csv')
transaction <- read.csv('data/transactions.csv')

# rename colname
head(account)
colnames(account)[1] <- 'account_id'
colnames(account)[3] <- 'open_date'

# join account with district name by district_id
head(account)
head(district)
account <- left_join(account, district[, 1:2], c("district_id"="id"))
colnames(account)[5] <- 'district_name'
account <- subset(account, select = -district_id)
head(account)

# join links and cards with client_id
link_card <- left_join(link, card, c("client_id"="link_id"))

# calculate the num_customers and join with account
num_customers <- link_card %>% group_by(account_id) %>% count(account_id, name = "num_customers") 
account<- left_join(account, num_customers)
head(account)

# calculate the credit_cards and join with account and set 0 for none
credit_cards <- link_card %>% group_by(account_id) %>% filter(!is.na(id.y)) %>% count(account_id, name = "credit_cards")
account<- left_join(account, credit_cards)
account[is.na(account$credit_cards),]$credit_cards <- 0

# create column that T/F if the loan is in default and join with account, set loan column for T/F whether the account has loan or not
loan$loan_default <- ifelse(loan$status == "B", T, F)
loan <- subset(loan, select = -c(date, id))
colnames(loan)[2:5] <- paste("loan", colnames(loan)[2:5], sep = "_")
account<- left_join(account, loan)
account$loan <- !is.na(account$loan_amount)


# group transaction data by account_id
transaction_acc <- transaction %>% group_by(account_id)
# get max and min balance data, withdrawal data and count of credit payments
max_balance = max(balance)
min_balance = min(balance)
balance <- transaction_acc %>% summarise(max_balance, min_balance)

withdrawal <- transaction_acc %>% filter(type == "debit") %>% summarise(max_withdrawal = max(amount), min_withdrawal = min(amount))
payment <- transaction_acc %>% filter(method == "credit card") %>% count(account_id, name="cc_payments")

# for join_all function. it cannot be library at the beginning otherwise count function can have errors
library(plyr)

# join all variables
analytical <- join_all(list(account, withdrawal, payment, balance), by = 'account_id', type = 'left')
analytical <- analytical[,c(1,4,2:3,5:6,12,7:11, 13:17)]

write.csv(analytical, file = "analytical_r.csv", row.names = FALSE)

File: analytical.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Jan 30 19:59:54 2022

@author: jieyisun
"""

import pandas as pd
import numpy as np

account = pd.read_csv('accounts.csv')
card = pd.read_csv('cards.csv')
loan = pd.read_csv('loans_py.csv')
district = pd.read_csv('districts.csv')
link = pd.read_csv('links.csv')
transaction = pd.read_csv('transactions.csv')

account_district = pd.merge(account, district.iloc[:, 0:2], left_on='district_id', right_on='id', how='left')
account_district = account_district.rename(columns={'id_x': 'account_id', 'date': 'open_date', 'name': 'district_name'}).drop(['district_id', 'id_y'], axis=1)

link_card = pd.merge(link, card, left_on='client_id', right_on='link_id', how='left')
num_customers = link_card.groupby(['account_id'])['client_id'].count().reset_index(name='num_customers')
credit_cards = link_card.groupby(['account_id'])['id_y'].count().reset_index(name='credit_cards')

df1 = pd.merge(pd.merge(account_district, num_customers, on='account_id', how='left'), credit_cards, on='account_id', how='left')
loan = loan.drop(['id', 'date'], axis=1)
loan.columns = 'loan_' + loan.columns
loan['loan_default'] = np.where((loan['loan_status'] == 'B'), True, False)

df2 = pd.merge(df1, loan, left_on='account_id', right_on='loan_account_id', how='left').drop('loan_account_id', axis=1)

df2['loan'] = np.where((df2["loan_amount"].isnull().values == True), False, True)

max_withdrawal = transaction[transaction.type == "debit"].groupby(['account_id'])['amount'].max().reset_index(name='max_withdrawal')
min_withdrawal = transaction[transaction.type == "debit"].groupby(['account_id'])['amount'].min().reset_index(name='min_withdrawal')
df3 = pd.merge(df2, pd.merge(max_withdrawal, min_withdrawal, on="account_id", how="outer"), on="account_id", how="outer")

cc_payment = transaction[transaction.method == "credit card"].groupby(['account_id'])['method'].count().reset_index(name='cc_payments')
df4 = pd.merge(df3, cc_payment, on='account_id', how='left')

max_balance = transaction.groupby(['account_id'])['balance'].max().reset_index(name='max_balance')
min_balance = transaction.groupby(['account_id'])['balance'].min().reset_index(name='min_balance')
df5 = pd.merge(df4, pd.merge(max_balance, min_balance, on="account_id", how="outer"), on="account_id", how="outer")

df5.to_csv("analytical_py.csv", index=0)
